### ISSUE TITLE
RowMajor float4x3 matrix in uniform buffer generates incorrect MSL

### ISSUE THREAD

**Issue by danginsburg** [KhronosGroup/SPIRV-Cross#436]
I have a shader that has two RowMajor float4x3 matrices in a uniform buffer:

```
               OpMemberDecorate %_struct_1365 0 RowMajor
               OpMemberDecorate %_struct_1365 0 Offset 0
               OpMemberDecorate %_struct_1365 0 MatrixStride 16
               OpMemberDecorate %_struct_1365 1 RowMajor
               OpMemberDecorate %_struct_1365 1 Offset 48
               OpMemberDecorate %_struct_1365 1 MatrixStride 16
               OpDecorate %_struct_1365 Block
```

The MSL output for this cbuffer is the following (./spirv-cross --output sky_vs.msl sky_vs.spv --msl):

```
struct _1365
{
    float4x3 _m0;
    float4x3 _m1;
};
```

This results in rendering corruption because according to the Metal Shading Language spec Table 5 a float4x3 is 64 bytes whereas these are supposed to be 48 bytes (a float3x4 in MSL).  It seems like spirv-cross would need to generate these as flaot3x4 in the struct definition and then transform them at point of use?  (unless there is some other way to specify row major that I'm not aware of in MSL).

I notice that the GLSL looks like it's probably correct:

```
layout(binding = 13, std140) uniform _1365_18812
{
    layout(row_major) mat4x3 _m0;
    layout(row_major) mat4x3 _m1;
} _18812;
```

The offending shader is attached to the bug.

Seems possibly related to #295.

[sky_vs.zip](https://github.com/KhronosGroup/SPIRV-Cross/files/1699470/sky_vs.zip)

**Comment by billhollings** [comment]
@danginsburg 

As far as I can tell...in SPIR-V, the matrices are 4 columns and 3 rows (4 columns of vec3)...so float4x3 is the correct declaration...but is the wrong memory footprint.

While float3x4 would have the right memory footprint...it wouldn't be the correct shape for use in the shader. We'd have to convert it to float4x3 dynamically in code.

Current conversions between R-Maj & C-Maj are working...but this issue would also require us to declare the wrong matrix shape and then transform it on the fly whenever we used it. We'd have to track something other than whether it was R-Maj or C-Maj.

Still thinking about this.

**Comment by danginsburg** [comment]
@billhollings Yeah, that's pretty much what I had in mind.  I'm not sure if it works, but an alternative may also be using a packed_float3 arrName[4]?  If I understand what that does it seems like it would be the right memory layout.  You'd still have to convert it to a float4x3 at use though.  This shader already is doing a conversion at use with spvConvertFromRowMajor4x3 so maybe it's a matter of doing that with an argument that takes a packed_float3 arr[4] in the case of a RowMajor float4x3?

I'm not sure which declaration is better (float3x4 vs packed_float3 arr[4]) or if it makes any difference.  Either way though, it seems doing a conversion at use is required.

**Comment by HansKristian-ARM** [comment]
If MSL has no row_major layout, it will be necessary to declare the matrix in the wrong order. Transposition will flip the column/row counts.

It is also possible to get an implied transposition by flipping multiplication order to avoid costly transpose(). HLSL backend basically does this.

**Comment by billhollings** [comment]
I'm working on this now.

**Comment by billhollings** [comment]
Fixed in PR #448.

**Comment by danginsburg** [comment]
I tested PR #448 and can confirm it fixes the rendering issue I was seeing.  Thanks!
